### PR TITLE
Cwhub testing

### PR DESCRIPTION
**Pull request by sbs2001**
related to https://github.com/crowdsecurity/crowdsec/pull/1434 

 - Fix collection upgrade mechanism
 - Tests :
  - [x] Download index, install collection
  - [x] Download index, install collection. Add scenario to collection (hub-side), update index, upgrade collection
  - [x] Download index, install collection. Locally remove scenario, update index, upgrade collection
  - [x] download index, install collection. Locally remove scenario, add scenario to collection (hub-side), update index, upgrade collection

### CODE REVIEW

File: pkg/cwhub/helpers_test.go [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+	require.True(t, hubIdx[COLLECTIONS]["crowdsecurity/test_collection"].Downloaded)
+	require.True(t, hubIdx[COLLECTIONS]["crowdsecurity/test_collection"].Installed)
+	require.False(t, hubIdx[COLLECTIONS]["crowdsecurity/test_collection"].UpToDate)
+	require.False(t, hubIdx[COLLECTIONS]["crowdsecurity/test_collection"].Tainted)
+
+	UpgradeConfig(cfg, COLLECTIONS, "crowdsecurity/test_collection", false)
+	assertCollectionDepsInstalled(t, "crowdsecurity/test_collection")
+
+	require.True(t, hubIdx[SCENARIOS]["crowdsecurity/barfoo_scenario"].Downloaded)
+	require.True(t, hubIdx[SCENARIOS]["crowdsecurity/barfoo_scenario"].Installed)
```
Comment (sbs2001) [inline pkg/cwhub/helpers_test.go:53]: This check fails in the context of spring4shell issue

File: pkg/cwhub/helpers.go [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+		return fmt.Errorf("unable to retrieve item : %s", name)
+	}
+	item := *it
+	if downloadOnly && item.Downloaded && item.UpToDate {
+		log.Warningf("%s is already downloaded and up-to-date", item.Name)
+		if !force {
+			return nil
+		}
+	}
+	item, err := DownloadLatest(csConfig.Hub, item, force, false)
```
Comment (sbs2001) [inline pkg/cwhub/helpers.go:23]:
```suggestion
item, err := DownloadLatest(csConfig.Hub, item, force, true)
```
Doing so will fix the test `TestUpgradeConfigNewScenarioInCollection`

File: pkg/cwhub/download.go [hunk 3]
```diff
@@ -90,10 +90,11 @@ func DownloadLatest(hub *csconfig.Hub, target Item, overwrite bool, updateOnly b
 					return target, fmt.Errorf("required %s %s of %s doesn't exist, abort", ptrtype, p, target.Name)
 				}
 
-				if !val.Installed && updateOnly {
+				if !val.Installed && updateOnly && val.Downloaded {
```
Comment (sbs2001) [inline pkg/cwhub/download.go:93]: val.Downloaded check does fixes the download of disabled scenario

File: pkg/cwhub/loader.go [hunk 4]
```diff
@@ -231,11 +234,13 @@ func parser_visit(path string, f os.FileInfo, err error) error {
...
-		} else if !inhub {
-			hubIdx[ftype][k] = v
-		}
+		// if _, ok := hubIdx[ftype][k]; !ok || !inhub || v.D {
+		// 	fmt.Printf("Updating %s", k)
+		// 	hubIdx[ftype][k] = v
+		// } else if !inhub {
+
+		// } else if
+		hubIdx[ftype][k] = v
```
Comment (sbs2001) [inline pkg/cwhub/loader.go:243]: This is overwrite for now. We don't know full consequences of this. As of now tests seem to pass

**Review by AlteredCoder — APPROVED** [review]

**Pull request merged**